clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1238,6 +1376,175 @@ namespace {
+        ProofResult Result =
+            CompareLowerOffsetsImpl(R, Cause, EquivExprs, Facts);
+        // If we get a definite result, no need to check free variables.
+        if (Result != ProofResult::Maybe)
+          return Result;
+
+        FreeVariablePosition DeclaredLowerPos = CombineFreeVariablePosition(
+            FreeVariablePosition::Declared, FreeVariablePosition::Lower);
+        FreeVariablePosition ObservedLowerPos = CombineFreeVariablePosition(
+            FreeVariablePosition::Observed, FreeVariablePosition::Lower);
+
+        if (CheckFreeVarInExprs(S, R.LowerOffsetVariable, LowerOffsetVariable,
+                                DeclaredLowerPos, ObservedLowerPos,
+                                EquivExprs, FreeVariables)) {
+          Cause = CombineFailures(Cause, ProofFailure::HasFreeVariables);
+          return ProofResult::False;
+        }
+
+        return ProofResult::Maybe;
+      }
+
+      // CompareUpperOffsets first calls CompareUpperOffsetsImpl. If
+      // CompareUpperOffsetsImpl returns Maybe, it continues to collect free
+      // variables in the upper offsets of this and R.
+      ProofResult CompareUpperOffsets(BaseRange &R, ProofFailure &Cause,
+                                      EquivExprSets *EquivExprs,
+                                      std::pair<ComparisonSet, ComparisonSet> &Facts,
+                                      FreeVariableListTy &FreeVariables) {
+        ProofResult Result =
+            CompareUpperOffsetsImpl(R, Cause, EquivExprs, Facts);
+        // If we get a definite result, no need to check free variables.
+        if (Result != ProofResult::Maybe || !EquivExprs)
+          return Result;
+
+        FreeVariablePosition DeclaredUpperPos = CombineFreeVariablePosition(
+            FreeVariablePosition::Declared, FreeVariablePosition::Upper);
+        FreeVariablePosition ObservedUpperPos = CombineFreeVariablePosition(
+            FreeVariablePosition::Observed, FreeVariablePosition::Upper);
+
+        if (CheckFreeVarInExprs(S, R.UpperOffsetVariable, UpperOffsetVariable,
+                                DeclaredUpperPos, ObservedUpperPos,
+                                EquivExprs, FreeVariables)) {
+          Cause = CombineFailures(Cause, ProofFailure::HasFreeVariables);
+          return ProofResult::False;
+        }
+
+        return ProofResult::Maybe;
+      }
+
+      // GetFreeVariables gathers "free variables" in SrcVars.
+      //
+      // Given two variable sets SrcVars and DstVars, and a set of equivalent
+      // sets of Expr EquivExprs. A variable V in SrcVars is *free* if the two
+      // conditions are met:
+      //   1. either (1) V is not an integer type, or (2) there is no set in
+      //   EquivExprs that contains both V and a constant (i.e. an integer-typed
+      //   value); and
+      //   2. for each variable U in DstVars, there is not set in EquivExprs
+      //   that contains both V and U.
```

Comment:
I think condition 2 needs to be amended to work in terms of uses of variables.   There is not a set in EquivExprs that contains two different expressions, where one expression uses V and the other expression uses U.